lib/membrane_wav/parser.ex

```diff
@@ -0,0 +1,243 @@
+     |      "data"       |    data length    |                 DATA                  |
+     |                   |     in bytes      |                                       |
+     |___________________|___________________|_______________________________________|
+    ```
+    Header may contain additional bytes between `bits per sample` and `"fact"` in case of `format`
+    different from 1 (1 represents PCM / uncompressed format). Length of block from `format` until
+    `"fact"` is present in `format block length` (it is 16 for PCM).
+
+    Blocks from byte 36 to 48 are optional. There can be additional bytes after `samples per
+    channel` if `fact block length` contains number bigger than 4.
+
+    Stages of parsing:
+    - `:init` - Parser waits for the first 22 bytes. After getting them, it parses these bytes
+      to ensure that it is a WAV file. Parser knows `format block length` and `format`, so it
+      is able to raise an error in case of different `format` than 1 (PCM) or different
+      length than 16 (for PCM). After parsing, the stage is set to `:format`.
+    - `:format` - Parser waits for the next 22 bytes - `fmt` subchunk (bytes 20 - 35) without
+      `format` and either `"fact"` and `fact block length` or `"data"` and `data length in bytes`.
+      Then it parses it and create `Membrane.Caps.Audio.Raw` struct with audio format to send it
+      as caps to the next element. Stage is set to `:fact` or `:data` depending on last 8 bytes.
+    - `:fact` - Parser waits for `8 + fact block length` bytes. It  parses them only to check if
+      the header is correct, but does not use that data in any way. After parsing, the stage is
+      set to `:data`.
+    - `:data` - header is already fully parsed. All new data from the input is sent to the output.
+  """
+
+  use Membrane.Filter
+
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
+  alias Membrane.Caps.Audio.Raw.Format
+
+  require Membrane.Logger
+
+  @pcm_format_size 16
+
+  @init_stage_size 22
+  @format_stage_size 22
+  @fact_stage_base_size 8
+
+  def_options frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :always,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(state) do
+    state = Map.merge(state, %{stage: :init, queue: <<>>})
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_start_of_stream(:input, _context, state) do
+    demand = {:input, @init_stage_size}
+
+    {{:ok, demand: demand}, state}
```

Comment:
I think you need to demand earlier than that, you won't get `start_of_stream` unless you demand something

Reply:
So maybe in `handle_prepared_to_playing/2`?

Reply:
Seems like the right place :+1: